tree/ntuple/v7/inc/ROOT/RDaos.hxx

```diff
@@ -167,25 +181,40 @@ private:
      \param vec A `std::vector<RWOperation>` that describes read/write operations to perform.
      \param cid The `daos_oclass_id_t` used to qualify OIDs.
      \param fn Either `std::mem_fn<&RDaosObject::Fetch>` (read) or `std::mem_fn<&RDaosObject::Update>` (write).
-     \return Number of requests that did not complete; this should be 0 after a successful call.
+     \return Error code (< 0) in case of failure, 0 on success.
      */
    template <typename Fn>
    int VectorReadWrite(std::vector<RWOperation> &vec, ObjClassId_t cid, Fn fn) {
+      using request_t = std::tuple<std::unique_ptr<RDaosObject>, RDaosObject::FetchUpdateArgs>;
+
       int ret;
-      DaosEventQueue eventQueue(vec.size());
-      {
-         std::vector<std::tuple<std::unique_ptr<RDaosObject>, RDaosObject::FetchUpdateArgs>> requests{};
-         requests.reserve(vec.size());
-         for (size_t i = 0; i < vec.size(); ++i) {
-           requests.push_back(std::make_tuple(std::unique_ptr<RDaosObject>(new RDaosObject(*this, vec[i].fOid, cid.fCid)),
-                                               RDaosObject::FetchUpdateArgs{
-                                                 vec[i].fDistributionKey, vec[i].fAttributeKey,
-                                                 vec[i].fIovs, &eventQueue.fEvs[i]}));
-            fn(std::get<0>(requests.back()).get(), std::get<1>(requests.back()));
-         }
-         ret = eventQueue.Poll();
+      std::vector<request_t> requests{};
+      requests.reserve(vec.size());
+
+      // Initialize symbolic parent event for all requests
+      daos_event_t parent_event{};
+      if ((ret = fPool->fEventQueue.InitializeEvent(&parent_event)) < 0)
+         return ret;
+
+      for (size_t i = 0; i < vec.size(); ++i) {
+         requests.push_back(std::make_tuple(
+            std::make_unique<RDaosObject>(*this, vec[i].fOid, cid.fCid),
+            RDaosObject::FetchUpdateArgs{vec[i].fDistributionKey, vec[i].fAttributeKey, vec[i].fIovs, true}));
+
+         // Initialize request event under parent
+         if ((ret = fPool->fEventQueue.InitializeEvent(&(std::get<1>(requests.back()).fEvent), &parent_event)) < 0)
+            return ret;
+
+         // Launch operation
+         if ((ret = fn(std::get<0>(requests.back()).get(), std::get<1>(requests.back()))) < 0)
+            return ret;
       }
-      return ret;
+
+      if ((ret = fPool->fEventQueue.LaunchParentBarrier(&parent_event)) < 0)
+         return ret;
+
+      // Poll parent until completion of all children launched before the barrier
+      return fPool->fEventQueue.PollEvent(&parent_event);
```

Comment:
As said in a previous comment, it might make sense to merge `LaunchParentBarrier()` and `PollEvent()`.
However, the current implementation of `PollEvent()` also finalizes the event, which is non-symmetrical with the call to `InitializeEvent()` on the parent - actually, on a first sight I though that the parent event was not ever released.

Reply:
Good point, I have decoupled `FinalizeEvent()` from the event polling functionality.